File: Task1.py
# Function to calculate the discount based on rules
def calculate_discount(cart_total, product_quantities):
    discount_rules = {
        "flat_10_discount": (cart_total > 200, 10),
        "bulk_5_discount": any(qty > 10 for qty in product_quantities),
        "bulk_10_discount": sum(product_quantities) > 20,
        "tiered_50_discount": sum(product_quantities) > 30 and any(qty > 15 for qty in product_quantities)
    }

    applicable_discounts = {rule: discount for rule, (condition, discount) in discount_rules.items() if condition}
    if applicable_discounts:
        max_discount_rule = max(applicable_discounts, key=applicable_discounts.get)
        return max_discount_rule, applicable_discounts[max_discount_rule]
    else:
        return None, 0

# Function to calculate shipping fee
def calculate_shipping_fee(total_units):
    return (total_units // 10) * 5

# Function to calculate gift wrap fee
def calculate_gift_wrap_fee(product_quantities):
    return sum(product_quantities)

# Function to calculate the total cost
def calculate_total_cost(product_quantities, product_prices):
    cart_total = sum(qty * price for qty, price in zip(product_quantities, product_prices))
    discount_rule, discount_amount = calculate_discount(cart_total, product_quantities)
    shipping_fee = calculate_shipping_fee(sum(product_quantities))
    gift_wrap_fee = calculate_gift_wrap_fee(product_quantities)

    discounted_total = cart_total - discount_amount
    total_cost = discounted_total + shipping_fee + gift_wrap_fee

    return cart_total, discount_rule, discount_amount, shipping_fee, gift_wrap_fee, total_cost

# Input product quantities and gift wrap information
product_quantities = []
product_prices = [20, 40, 50]  # Prices for Product A, B, C

for product in ["A", "B", "C"]:
    quantity = int(input(f"Enter the quantity of Product {product}: "))
    product_quantities.append(quantity)

    gift_wrap = input(f"Is Product {product} wrapped as a gift? (yes/no): ").lower()
    if gift_wrap == "yes":
        product_quantities[-1] += 1  # Increment quantity for gift wrap

# Calculate and display the details
subtotal, discount_rule, discount_amount, shipping_fee, gift_wrap_fee, total = calculate_total_cost(product_quantities, product_prices)

print("\nInvoice Details:")
for i, product in enumerate(["A", "B", "C"]):
    print(f"Product {product}: Quantity - {product_quantities[i]}, Total Amount - ${product_quantities[i] * product_prices[i]}")

print(f"\nSubtotal: ${subtotal}")
print(f"Discount Applied ({discount_rule}): -${discount_amount}")
print(f"Shipping Fee: ${shipping_fee}")
print(f"Gift Wrap Fee: ${gift_wrap_fee}")
print(f"Total: ${total}")

File: Task1.js
// Function to calculate the discount based on rules
function calculateDiscount(cartTotal, productQuantities) 
{
    const discountRules = 
    {
        flat_10_discount: cartTotal > 200 ? 10 : 0,
        bulk_5_discount: productQuantities.some(qty => qty > 10) ? 5 : 0,
        bulk_10_discount: productQuantities.reduce((sum, qty) => sum + qty, 0) > 20 ? 10 : 0,
        tiered_50_discount: productQuantities.reduce((sum, qty) => sum + qty, 0) > 30 && productQuantities.some(qty => qty > 15) ? 50 : 0
    };

    const applicableDiscounts = Object.entries(discountRules).filter(([rule, discount]) => discount > 0);
    if (applicableDiscounts.length > 0) 
    {
        const [maxDiscountRule, maxDiscount] = applicableDiscounts.reduce((max, curr) => (curr[1] > max[1] ? curr : max));
        return [maxDiscountRule, maxDiscount];
    } 
    else 
    {
        return [null, 0];
    }
}

// Function to calculate shipping fee
function calculateShippingFee(totalUnits) 
{
    return Math.floor(totalUnits / 10) * 5;
}

// Function to calculate gift wrap fee
function calculateGiftWrapFee(productQuantities) 
{
    return productQuantities.reduce((sum, qty) => sum + qty, 0);
}

// Function to calculate the total cost
function calculateTotalCost(productQuantities, productPrices)
 {
    const cartTotal = productQuantities.reduce((sum, qty, index) => sum + qty * productPrices[index], 0);
    const [discountRule, discountAmount] = calculateDiscount(cartTotal, productQuantities);
    const shippingFee = calculateShippingFee(productQuantities.reduce((sum, qty) => sum + qty, 0));
    const giftWrapFee = calculateGiftWrapFee(productQuantities);

    const discountedTotal = cartTotal - discountAmount;
    const totalCost = discountedTotal + shippingFee + giftWrapFee;

    return [cartTotal, discountRule, discountAmount, shippingFee, giftWrapFee, totalCost];
}

// Input product quantities and gift wrap information
const productQuantities = [];
const productPrices = [20, 40, 50]; // Prices for Product A, B, C

for (let i = 0; i < 3; i++) 
{
    const quantity = parseInt(prompt(`Enter the quantity of Product ${String.fromCharCode(65 + i)}: `), 10);
    productQuantities.push(quantity);

    const giftWrap = prompt(`Is Product ${String.fromCharCode(65 + i)} wrapped as a gift? (yes/no): `).toLowerCase();
    if (giftWrap === "yes") {
        productQuantities[i]++;
    }
}

// Calculate and display the details
const [subtotal, discountRule, discountAmount, shippingFee, giftWrapFee, total] = calculateTotalCost(productQuantities, productPrices);

console.log("\nInvoice Details:");
["A", "B", "C"].forEach((product, i) => 
{
    console.log(`Product ${product}: Quantity - ${productQuantities[i]}, Total Amount - $${productQuantities[i] * productPrices[i]}`);
});

console.log(`\nSubtotal: $${subtotal}`);
console.log(`Discount Applied (${discountRule}): -$${discountAmount}`);
console.log(`Shipping Fee: $${shippingFee}`);
console.log(`Gift Wrap Fee: $${giftWrapFee}`);
console.log(`Total: $${total}`);
